src/System.Linq/src/System/Linq/Intersect.cs

```diff
@@ -19,8 +19,17 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
             {
                 throw Error.ArgumentNull(nameof(second));
             }
+           
+            var first2 = first.Distinct();
+            var second2 = second.Distinct();
```

Comment:
The current code creates a `Set<T>` which has some up-front cost, and then removes from it on the fly. This creates `1 + first.Count()` `Set<T>` so it has that up-front cost multiple times, and then loses the "knowledge" of the set so `Contains` can't be optimised. Did you re-try the benchmarks when you changed the code?